lib/rucio/tests/test_download.py

```diff
@@ -463,3 +464,17 @@ def __init__(self, *args, mocks_get=mocks_get, **kwargs):
             mocks_get.clear()
             download_client.download_dids([{'did': did_str, 'base_dir': tmp_dir, 'transfer_speed_timeout': 0}])
             mocks_get[0].assert_called_with(ANY, ANY, transfer_timeout=60)
+
+
+def test_download_file_with_impl(rse_factory, did_factory, download_client):
+    impl = 'rucio.rse.protocols.xrootd.Default' 
+    scope = str(did_factory.default_scope)
+    rse, _ = rse_factory.make_posix_rse()
+    did = did_factory.upload_test_file(rse)
+    did_str = '%s:%s' % (did['scope'], did['name'])
+    
+    #checks if the protocol object returned has the same impl value as what was specified, thus in turn implying the right module was called
+    result = download_client.download_dids([{'did': did_str, 'impl': impl}], test_module=True)
+    for x in range(len(result)):
+        print(result[x])
```

Comment:
this was for me to see what was actually being passed, and this is when I realised why an error kept coming up:
```
[root@b2e68ef33f26 rucio]# tools/pytest.sh test_download.py
Running pytest with extra arguments: test_download.py
Running pytest with pytest-xdist: -p xdist -p ruciopytest.plugin --dist=rucio --numprocesses=auto
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.6.8, pytest-6.1.2, py-1.10.0, pluggy-0.13.1
rootdir: /opt/rucio/lib/rucio/tests
plugins: xdist-2.2.1
gw0 [10] / gw1 [10] / gw2 [10] / gw3 [10] / gw4 [10] / gw5 [10] / gw6 [10] / gw7 [10]
s.F......
==================================================================================================== FAILURES =====================================================================================================
__________________________________________________________________________________________ test_download_file_with_impl ___________________________________________________________________________________________
[gw3] linux -- Python 3.6.8 /usr/bin/python

rse_factory = <rucio.tests.temp_factories.TemporaryRSEFactory object at 0x7f6482fbe518>, did_factory = <rucio.tests.temp_factories.TemporaryDidFactory object at 0x7f6482fbe898>
download_client = <rucio.client.downloadclient.DownloadClient object at 0x7f6482fccb38>

    def test_download_file_with_impl(rse_factory, did_factory, download_client):
        impl = 'rucio.rse.protocols.xrootd.Default'
        scope = str(did_factory.default_scope)
        rse, _ = rse_factory.make_posix_rse()
        did = did_factory.upload_test_file(rse)
        did_str = '%s:%s' % (did['scope'], did['name'])
    
        #checks if the protocol object returned has the same impl value as what was specified, thus in turn implying the right module was called
        result = download_client.download_dids([{'did': did_str, 'impl': impl}], test_module=True)
        for x in range(len(result)):
            print(result[x])
>       assert result.test_impl() == impl
E       AttributeError: 'list' object has no attribute 'test_impl'

test_download.py:480: AttributeError
---------------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------------
<rucio.rse.protocols.posix.Default object at 0x7f64807a8a90>
---------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------
Processing 1 item(s) for input
num_unmerged_items=1; num_dids=1; num_merged_items=1
Getting sources of DIDs
No tapes found.
schemes: None
rse_expression: None
num DIDs for list_replicas call: 1
num resolved files: 1
"unzip -v" returned with exitcode 127
"tar --version" returned with exitcode 0
Queueing file: mock:file_TINRAJFCLL
real parents: set()
options: {'mock:file_TINRAJFCLL': {'destinations': {('.', False)}, 'ignore_checksum': False, 'transfer_timeout': None, 'transfer_speed_timeout': None}}
Prepared sources: num_sources=1/1; num_non_cea_sources=1; num_cea_ids=0
Using main thread to download 1 file(s)
Start processing queued downloads
Preparing download of mock:file_TINRAJFCLL
Trying to download with file and timeout of 360s from MOCK_UJJZPBZBDR: mock:file_TINRAJFCLL 
------------------------------------------------------------------------------------------------ Captured log call ------------------------------------------------------------------------------------------------
WARNING  root:uploadclient.py:368 Scope mock not found for the account root.
INFO     dlul_client:downloadclient.py:278 Processing 1 item(s) for input
DEBUG    dlul_client:downloadclient.py:281 num_unmerged_items=1; num_dids=1; num_merged_items=1
INFO     dlul_client:downloadclient.py:283 Getting sources of DIDs
DEBUG    dlul_client:downloadclient.py:1119 No tapes found.
DEBUG    dlul_client:downloadclient.py:1126 schemes: None
DEBUG    dlul_client:downloadclient.py:1130 rse_expression: None
DEBUG    dlul_client:downloadclient.py:1133 num DIDs for list_replicas call: 1
DEBUG    dlul_client:downloadclient.py:1145 num resolved files: 1
DEBUG    dlul_client:downloadclient.py:86 "unzip -v" returned with exitcode 127
DEBUG    dlul_client:downloadclient.py:86 "tar --version" returned with exitcode 0
DEBUG    dlul_client:downloadclient.py:1228 Queueing file: mock:file_TINRAJFCLL
DEBUG    dlul_client:downloadclient.py:1229 real parents: set()
DEBUG    dlul_client:downloadclient.py:1230 options: {'mock:file_TINRAJFCLL': {'destinations': {('.', False)}, 'ignore_checksum': False, 'transfer_timeout': None, 'transfer_speed_timeout': None}}
DEBUG    dlul_client:downloadclient.py:1309 Prepared sources: num_sources=1/1; num_non_cea_sources=1; num_cea_ids=0
INFO     dlul_client:downloadclient.py:368 Using main thread to download 1 file(s)
DEBUG    dlul_client:downloadclient.py:413 Start processing queued downloads
INFO     dlul_client:downloadclient.py:479 Preparing download of mock:file_TINRAJFCLL
INFO     dlul_client:downloadclient.py:590 Trying to download with file and timeout of 360s from MOCK_UJJZPBZBDR: mock:file_TINRAJFCLL
================================================================================================ warnings summary =================================================================================================
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
  /opt/rucio/lib/rucio/client/baseclient.py:55: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

test_download.py::test_download_blacklisted_replicas
test_download.py::test_download_file_with_impl
test_download.py::test_download_without_base_dir
test_download.py::test_norandom_respected
test_download.py::test_trace_copy_out_and_checksum_validation
test_download.py::test_transfer_timeout
test_download.py::test_download_to_two_paths
test_download.py::test_download_multiple
  /usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py:1020: InsecureRequestWarning: Unverified HTTPS request is being made to host 'rucio'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
    InsecureRequestWarning,

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================= short test summary info =============================================================================================
FAILED test_download.py::test_download_file_with_impl - AttributeError: 'list' object has no attribute 'test_impl'
=============================================================================== 1 failed, 7 passed, 1 skipped, 16 warnings in 8.07s ===============================================================================
[root@b2e68ef33f26 rucio]# 

```

So particularly the stdout log shows: <rucio.rse.protocols.posix.Default object at 0x7f64807a8a90> is being passed back as obviously due to inner working it is considering the impl as the one associated with the RSE still and not what is being passed by the client, so anyways that is the feature I am implementing and thus this fail is expected right?
So is this test case okay? Most basic one